src/table_properties.rs

```diff
@@ -0,0 +1,300 @@
+
+use crate::Options;
+
+/// Extension trait for [`Options`] to register table properties collectors
+pub trait TablePropertiesExt {
+    fn add_table_properties_collector_factory<F>(&mut self, factory_fn: F)
+    where
+        F: TablePropertiesCollectorFactory + Send + 'static;
+}
+
+impl TablePropertiesExt for Options {
+    fn add_table_properties_collector_factory<F>(&mut self, factory: F)
+    where
+        F: TablePropertiesCollectorFactory + Send + 'static,
+    {
+        unsafe {
+            let raw_factory = Box::into_raw(Box::new(factory)) as *mut c_void;
+
+            // This method takes ownership of "factory" and transfers it to the FFI wrapper, which
+            // in turn will handed over to RocksDB; we no longer track it after this point. This
+            // will have to change if we need to support removing collectors in the future.
+            ffi::rocksdb_options_add_table_properties_collector_factory(
+                self.inner,
+                raw_factory,
+                Some(TablePropertiesCollectorFactoryCallback::<F>::destructor),
+                Some(TablePropertiesCollectorFactoryCallback::<F>::get_name),
+                Some(TablePropertiesCollectorFactoryCallback::<F>::create_collector),
+            );
+        }
+    }
+}
+
+/// Table properties
+pub struct TableProperties<'a> {
+    // For now this struct exists to allow read-only access to user collected properties.
+    // Other fields will be exposed as needed.
+
+    // We own the underlying pointer wrapper struct
+    table_properties: *const rocksdb_table_properties_t,
+    _marker: PhantomData<&'a ()>,
+}
+
+impl Drop for TableProperties<'_> {
+    fn drop(&mut self) {
+        // Free the table properties holder struct
+        unsafe { ffi::rocksdb_table_properties_destroy(self.table_properties) };
+    }
+}
+
+impl<'a> TableProperties<'a> {
+    pub(crate) fn from_ptr(
+        table_properties_ptr: *const rocksdb_table_properties_t,
+    ) -> TableProperties<'a> {
+        TableProperties {
+            table_properties: table_properties_ptr,
+            _marker: PhantomData,
+        }
+    }
+
+    // This method wraps access to a custom user-collected property via a read-only
+    // pointer to the RocksDB FFI rocksdb_table_properties_t struct.
+    pub fn get_user_collected_property(&self, key: &str) -> Option<String> {
+        unsafe {
+            let user_collected_properties =
+                ffi::rocksdb_table_properties_get_user_collected_properties(self.table_properties);
+
+            let key_cstring = CString::new(key).unwrap();
+            let value_ptr = ffi::rocksdb_user_collected_properties_get(
+                user_collected_properties,
+                key_cstring.as_ptr(),
+            );
+
+            let value = if value_ptr.is_null() {
+                None
+            } else {
+                let value_string = CStr::from_ptr(value_ptr).to_str().unwrap().to_owned();
```

Comment:
Are you sure this is a utf-8 string?

Reply:
Absolutely not! I've updated the property collector interface to be `HashMap<CString, CString>` to make this clearer from both reading and writing these.